politeiad/backendv2/tstorebe/plugins/pi/cmds.go

```diff
@@ -0,0 +1,258 @@
+		Timestamp: time.Now().Unix(),
+		Receipt:   hex.EncodeToString(receipt[:]),
+	}
+	err = p.billingStatusSave(token, bsc)
+	if err != nil {
+		return "", err
+	}
+
+	// Prepare reply
+	sbsr := pi.SetBillingStatusReply{
+		Timestamp: bsc.Timestamp,
+		Receipt:   bsc.Receipt,
+	}
+	reply, err := json.Marshal(sbsr)
+	if err != nil {
+		return "", err
+	}
+
+	return string(reply), nil
+}
+
+// billingStatusSave saves a BillingStatusChange to the backend.
+func (p *piPlugin) billingStatusSave(token []byte, bsc pi.BillingStatusChange) error {
+	// Prepare blob
+	be, err := billingStatusEncode(bsc)
+	if err != nil {
+		return err
+	}
+
+	// Save blob
+	return p.tstore.BlobSave(token, *be)
+}
+
+// billingStatuses returns all BillingStatusChange for a record.
+func (p *piPlugin) billingStatuses(token []byte) ([]pi.BillingStatusChange, error) {
+	// Retrieve blobs
+	blobs, err := p.tstore.BlobsByDataDesc(token,
+		[]string{dataDescriptorBillingStatus})
+	if err != nil {
+		return nil, err
+	}
+
+	// Decode blobs
+	statuses := make([]pi.BillingStatusChange, 0, len(blobs))
+	for _, v := range blobs {
+		a, err := billingStatusDecode(v)
+		if err != nil {
+			return nil, err
+		}
+		statuses = append(statuses, *a)
+	}
+
+	// Sanity check. They should already be sorted from oldest to
+	// newest.
+	sort.SliceStable(statuses, func(i, j int) bool {
+		return statuses[i].Timestamp < statuses[j].Timestamp
+	})
+
+	return statuses, nil
+}
+
+// tokenMatches verifies that the command token (the token for the record that
+// this plugin command is being executed on) matches the payload token (the
+// token that the plugin command payload contains that is typically used in the
+// payload signature). The payload token must be the full length token.
+func tokenMatches(cmdToken []byte, payloadToken string) error {
+	pt, err := tokenDecode(payloadToken)
+	if err != nil {
+		return backend.PluginError{
+			PluginID:     pi.PluginID,
+			ErrorCode:    uint32(pi.ErrorCodeTokenInvalid),
+			ErrorContext: util.TokenRegexp(),
+		}
+	}
+	if !bytes.Equal(cmdToken, pt) {
+		return backend.PluginError{
+			PluginID:  pi.PluginID,
+			ErrorCode: uint32(pi.ErrorCodeTokenInvalid),
+			ErrorContext: fmt.Sprintf("payload token does not "+
+				"match command token: got %x, want %x",
+				pt, cmdToken),
+		}
+	}
+	return nil
+}
+
+// convertSignatureError converts a util SignatureError to a backend
+// PluginError that contains a pi plugin error code.
+func convertSignatureError(err error) backend.PluginError {
+	var e util.SignatureError
+	var s pi.ErrorCodeT
+	if errors.As(err, &e) {
+		switch e.ErrorCode {
+		case util.ErrorStatusPublicKeyInvalid:
+			s = pi.ErrorCodePublicKeyInvalid
+		case util.ErrorStatusSignatureInvalid:
+			s = pi.ErrorCodeSignatureInvalid
+		}
+	}
+	return backend.PluginError{
+		PluginID:     pi.PluginID,
+		ErrorCode:    uint32(s),
+		ErrorContext: e.ErrorContext,
+	}
+}
+
+func billingStatusEncode(bsc pi.BillingStatusChange) (*store.BlobEntry, error) {
```

Comment:
Would prefer these are listed up with the other `billingStatus` helper functions.  These also needs comments.